### PR TITLE
Stratified Splitter

### PR DESCRIPTION
**Pull request by mottodora**
- [x] Stratified Splitting for classification
- [x] Stratified Splitting for regression
- [x] Unit tests for classification
- [x] Unit tests for regression
- [x] fix seed
- [x] Documents

### CODE REVIEW

File: chainer_chemistry/dataset/splitters/stratified_splitter.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        labels_feature = dataset.features[:, labels_feature_id]
+        if len(labels_feature.shape) == 1:
+            labels = labels_feature
+        else:
+            labels = labels_feature[:, task_id]
+
+        if labels.dtype.kind == 'i':
+            classes, label_indices = numpy.unique(labels, return_inverse=True)
+        elif labels.dtype.kind == 'f':
+            n_bin = 10
```
Comment (corochann) [inline chainer_chemistry/dataset/splitters/stratified_splitter.py]: Can you add it as optional kwarg??

File: chainer_chemistry/dataset/splitters/stratified_splitter.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        if not isinstance(dataset, NumpyTupleDataset):
+            raise NotImplementedError
+        labels_feature = dataset.features[:, labels_feature_id]
+        if len(labels_feature.shape) == 1:
+            labels = labels_feature
+        else:
+            labels = labels_feature[:, task_id]
+
+        if labels.dtype.kind == 'i':
+            classes, label_indices = numpy.unique(labels, return_inverse=True)
```
[the remark this answers is not in the document]
Comment (mottodora) [reply chainer_chemistry/dataset/splitters/stratified_splitter.py]: `class_indices` is already used. I will use `labels`.

**Comment by mottodora** [comment]
resolve #206

File: chainer_chemistry/dataset/splitters/stratified_splitter.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+            if not isinstance(dataset, NumpyTupleDataset):
+                raise ValueError("Please assign label dataset.")
+            labels = dataset.features[:, label_axis]
+
+        if len(labels.shape) == 1:
+            labels = labels
+        else:
+            labels = labels[:, task_index]
+
+        if task_type == 'infer':
```
Comment (corochann) [inline chainer_chemistry/dataset/splitters/stratified_splitter.py]: auto

File: chainer_chemistry/dataset/splitters/stratified_splitter.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
+import numpy
+import pandas
+
+from chainer_chemistry.dataset.splitters.base_splitter import BaseSplitter
+from chainer_chemistry.datasets.numpy_tuple_dataset import NumpyTupleDataset
+
+
+def _approximate_mode(class_counts, n_draws):
```
Comment (corochann) [inline chainer_chemistry/dataset/splitters/stratified_splitter.py]: comment: url you referred.

File: chainer_chemistry/dataset/splitters/stratified_splitter.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+            classes, labels = numpy.unique(labels, return_inverse=True)
+        elif task_type == 'regression':
+            classes = numpy.arange(n_bin)
+            labels = pandas.qcut(labels, n_bin, labels=False)
+        else:
+            raise ValueError
+
+        n_classes = classes.shape[0]
+        n_total_valid = int(numpy.floor(frac_valid * len(dataset)))
+        n_total_test = int(numpy.floor(frac_test * len(dataset)))
```
Comment (corochann) [inline chainer_chemistry/dataset/splitters/stratified_splitter.py]: please put comment
\# n_total_train is the remainder: n - n_total_valid - n_total_test

File: chainer_chemistry/dataset/splitters/stratified_splitter.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+            train_index.extend(class_train_index)
+            valid_index.extend(class_valid_index)
+            test_index.extend(class_test_index)
+
+        assert n_total_valid == len(valid_index)
+        assert n_total_test == len(test_index)
+
+        return rng.permutation(train_index),\
+            rng.permutation(valid_index),\
+            rng.permutation(test_index),
```
Comment (corochann) [inline chainer_chemistry/dataset/splitters/stratified_splitter.py]: I think it is ok to just return array.

File: chainer_chemistry/dataset/splitters/stratified_splitter.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+        .. admonition:: Example
+            >>> from chainer_chemistry.datasets import NumpyTupleDataset
+            >>> from chainer_chemistry.dataset.splitters \
+            >>>                                 import StratifiedSplitter
+            >>> a = numpy.random.random((10, 10))
+            >>> b = numpy.random.random((10, 8))
+            >>> c = numpy.random.random((10, 1))
+            >>> d = NumpyTupleDataset(a, b, c)
+            >>> splitter = StratifiedSplitter()
+            >>> splitter.train_valid_test_split()
```
Comment (corochann) [inline chainer_chemistry/dataset/splitters/stratified_splitter.py]: remove this line

File: chainer_chemistry/dataset/splitters/stratified_splitter.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+        .. admonition:: Example
+            >>> from chainer_chemistry.datasets import NumpyTupleDataset
+            >>> from chainer_chemistry.dataset.splitters \
+            >>>                                 import StratifiedSplitter
+            >>> a = numpy.random.random((10, 10))
+            >>> b = numpy.random.random((10, 8))
+            >>> c = numpy.random.random((10, 1))
+            >>> d = NumpyTupleDataset(a, b, c)
+            >>> splitter = StratifiedSplitter()
+            >>> splitter.train_valid_split()
```
Comment (corochann) [inline chainer_chemistry/dataset/splitters/stratified_splitter.py]: remove this line

File: chainer_chemistry/dataset/splitters/stratified_splitter.py [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+                             "'regression' or 'infer'".format(task_type))
+
+        rng = numpy.random.RandomState(seed)
+
+        if labels is None:
+            if not isinstance(dataset, NumpyTupleDataset):
+                raise ValueError("Please assign label dataset.")
+            labels = dataset.features[:, label_axis]
+
+        if len(labels.shape) == 1:
```
Comment (corochann) [inline chainer_chemistry/dataset/splitters/stratified_splitter.py]: labels.ndim

File: chainer_chemistry/dataset/splitters/stratified_splitter.py [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+            raise ValueError("{} is invalid. Please use 'classification',"
+                             "'regression' or 'infer'".format(task_type))
+
+        rng = numpy.random.RandomState(seed)
+
+        if labels is None:
+            if not isinstance(dataset, NumpyTupleDataset):
+                raise ValueError("Please assign label dataset.")
+            labels = dataset.features[:, label_axis]
+
```
Comment (corochann) [inline chainer_chemistry/dataset/splitters/stratified_splitter.py]: if isinstance(labels, list):
    labels = numpy.array(labels)

File: chainer_chemistry/dataset/splitters/stratified_splitter.py (hunk 5 above)
Comment (mottodora) [reply chainer_chemistry/dataset/splitters/stratified_splitter.py]: I wrote comments in other place.

File: .pytest_cache/v/cache/nodeids [hunk 11]
```diff
@@ -0,0 +1,330 @@
+[
```
Comment (corochann) [inline .pytest_cache/v/cache/nodeids]: can you remove this file from commit?
and if not done, please put `.pytest_cache` to .gitignore.

File: tests/dataset_tests/splitters_tests/test_stratified_splitter.py [hunk 12]
```diff
@@ -52,7 +52,28 @@ def test_classification_split(cls_dataset):
...
+    train_ind, valid_ind, test_ind = splitter._split(cls_dataset,
+                                                     labels=cls_label,
+                                                     frac_train=0.5,
+                                                     frac_valid=0.3,
+                                                     frac_test=0.2)
+    assert type(train_ind) == numpy.ndarray
+    assert train_ind.shape[0] == 15
+    assert valid_ind.shape[0] == 9
+    assert test_ind.shape[0] == 6
+
```
Comment (corochann) [inline tests/dataset_tests/splitters_tests/test_stratified_splitter.py]: How about checking the number of class 0, 1 must be larget than some amount? to ensure stratified splitting?

File: tests/dataset_tests/splitters_tests/test_stratified_splitter.py [hunk 13]
```diff
@@ -52,7 +52,28 @@ def test_classification_split(cls_dataset):
...
+
+def test_classification_split_by_labels_list(cls_dataset, cls_label):
```
Comment (corochann) [inline tests/dataset_tests/splitters_tests/test_stratified_splitter.py]: How about checking the number of class 0, 1 must be larget than some amount? to ensure stratified splitting?

File: .pytest_cache/v/cache/nodeids (hunk 11 above)
Comment (mottodora) [reply .pytest_cache/v/cache/nodeids]: I already remove this file.

**Comment by mottodora** [comment]
Updated.

**Comment by corochann** [comment]
LGTM

**Pull request merged**